packages/web-client/tests/integration/components/card-space/create-space-workflow/username-test.ts

```diff
@@ -0,0 +1,249 @@
+      assert.dom(AVATAR).hasAttribute('src', sampleImage);
+      assert.dom(EDIT_BUTTON).containsText('Edit');
+
+      // assert that we can get back to an edit state with the appropriate information still there
+      await click(EDIT_BUTTON);
+      assert.dom(USERNAME_INPUT).hasValue('monchi');
+      assert.dom(IMAGE_UPLOADER_AVATAR).hasAttribute('src', sampleImage);
+      assert.dom(SAVE_BUTTON).exists();
+      assert.dom(EDIT_BUTTON).doesNotExist();
+    });
+
+    test('it prevents submission of an invalid username', async function (this: MirageTestContext, assert) {
+      this.server.post(
+        '/card-spaces/validate-profile-name',
+        function (_schema, request) {
+          let profileName = JSON.parse(request.requestBody).data.attributes[
+            'profile-name'
+          ];
+          if (profileName !== 'Errored') {
+            return {
+              errors: [],
+            };
+          } else {
+            return {
+              errors: [
+                {
+                  status: '422',
+                  title: 'Test invalid username',
+                  detail: 'Test invalid username',
+                },
+              ],
+            };
+          }
+        }
+      );
+
+      await render(hbs`
+        <CardSpace::CreateSpaceWorkflow::Username
+          @workflowSession={{this.session}}
+          @onComplete={{this.onComplete}}
+          @onIncomplete={{this.onIncomplete}}
+          @isComplete={{this.isComplete}}
+        />
+      `);
+
+      assert.dom(SAVE_BUTTON).isDisabled();
+
+      await fillIn(USERNAME_INPUT, 'Hello there');
+      assert.dom(USERNAME_INPUT).hasValue('Hello there');
+
+      await waitFor(
+        `${USERNAME_INPUT}[data-test-boxel-input-validation-state="valid"]`
+      );
+
+      assert.dom(SAVE_BUTTON).isEnabled();
+
+      await fillIn(USERNAME_INPUT, 'Errored');
+      assert.dom(USERNAME_INPUT).hasValue('Errored');
+
+      await waitFor(
+        `${USERNAME_INPUT}[data-test-boxel-input-validation-state="invalid"]`
+      );
+
+      assert.dom(SAVE_BUTTON).isDisabled();
+      assert.dom(USERNAME_FIELD).containsText('Test invalid username');
+    });
+
+    test('it can upload an image', async function (this: MirageTestContext, assert) {
+      this.server.namespace = '';
+      this.server.post('/upload', function () {
+        return {
+          data: {
+            type: 'uploaded-asset',
+            attributes: {
+              url: uploadResultImage,
+            },
+          },
+        };
+      });
+
+      await render(hbs`
+        <CardSpace::CreateSpaceWorkflow::Username
+          @workflowSession={{this.session}}
+          @onComplete={{this.onComplete}}
+          @onIncomplete={{this.onIncomplete}}
+          @isComplete={{this.isComplete}}
+        />
+      `);
+
+      // this opens the image editor. users would normally trigger this
+      // by clicking the image uploader
+      await mockPngUpload(sampleImage, IMAGE_UPLOADER);
```

Comment:
I appreciate that you put in the significant-seeming effort to make this look simple in the test! 🤩